src/datamodel/validator/UUri.cpp

```diff
@@ -10,3 +10,169 @@
+			       "form.";
+		default:
+			return "Unknown reason.";
+	}
+}
+
+bool uses_wildcards(const v1::UUri& uuri) {
+	if (uuri.authority_name().find("*") != std::string::npos) {
+		return true;
+	}
+	if ((uuri.ue_id() & 0xFFFF) == 0xFFFF) {  // service ID
+		return true;
+	}
+	if ((uuri.ue_id() & 0xFFFF0000) == 0) {  // service instance ID
+		return true;
+	}
+	if (uuri.ue_version_major() == 0xFF) {
+		return true;
+	}
+	if (uuri.resource_id() == 0xFFFF) {
+		return true;
+	}
+	return false;
+}
+
+ValidationResult isValid(const v1::UUri& uuri) {
+	auto [valid, reason] = isValidRpcMethod(uuri);
+	if (!valid) {
+		auto [valid, reason] = isValidRpcResponse(uuri);
+		if (!valid) {
+			auto [valid, reason] = isValidPublishTopic(uuri);
+			if (!valid) {
+				return isValidNotification(uuri);
+			}
+		}
+	}
+
+	return {true, std::nullopt};
+}
+
+ValidationResult isValidRpcMethod(const v1::UUri& uuri) {
+	if (uuri.authority_name().empty()) {
+		return {false, Reason::EMPTY};
+	}
+
+	// disallow wildcards
+	if (uses_wildcards(uuri)) {
+		return {false, Reason::DISALLOWED_WILDCARD};
+	}
+
+	// check resource ID [0x0001, 0x7FFF]
+	uint16_t resource_id = uuri.resource_id() & 0xFFFF;
+	if (resource_id == 0 || resource_id > 0x7FFF) {
+		return {false, Reason::BAD_RESOURCE_ID};
+	}
+
+	return {true, std::nullopt};
+}
+
+ValidationResult isValidRpcResponse(const v1::UUri& uuri) {
+	if (uuri.authority_name().empty()) {
+		return {false, Reason::EMPTY};
+	}
+
+	// disallow wildcards
+	if (uses_wildcards(uuri)) {
+		return {false, Reason::DISALLOWED_WILDCARD};
+	}
+
+	if (uuri.resource_id() != 0) {
+		return {false, Reason::BAD_RESOURCE_ID};
+	}
+	return {true, std::nullopt};
+}
+
+ValidationResult isValidDefaultSource(const v1::UUri& uuri) {
+	return isValidRpcResponse(uuri);
+}
+
+ValidationResult isValidPublishTopic(const v1::UUri& uuri) {
+	if (uuri.authority_name().empty()) {
+		return {false, Reason::EMPTY};
+	}
+
+	// disallow wildcards
+	if (uses_wildcards(uuri)) {
+		return {false, Reason::DISALLOWED_WILDCARD};
+	}
+
+	if ((uuri.resource_id() & 0xFFFF) < 0x8000) {
+		return {false, Reason::BAD_RESOURCE_ID};
+	}
+
+	return {true, std::nullopt};
+}
+
+ValidationResult isValidNotification(const v1::UUri& uuri) {
+	if (uuri.authority_name().empty()) {
+		return {false, Reason::EMPTY};
+	}
+
+	// disallow wildcards
+	if (uses_wildcards(uuri)) {
+		return {false, Reason::DISALLOWED_WILDCARD};
+	}
+
+	if ((uuri.resource_id() & 0xFFFF) < 0x8000) {
+		return {false, Reason::BAD_RESOURCE_ID};
+	}
+
+	return {true, std::nullopt};
+}
+
+ValidationResult isValidSubscription(const v1::UUri& uuri) {
+	if (uuri.authority_name().empty()) {
+		return {false, Reason::EMPTY};
+	}
+
+	if ((uuri.resource_id() & 0xFFFF) < 0x8000) {
+		return {false, Reason::BAD_RESOURCE_ID};
+	}
+
+	return isValidPublishTopic(uuri);
+}
+
+ValidationResult isEmpty(const v1::UUri& uuri) {
+	if (uuri.authority_name().find_first_not_of(" ") != std::string::npos) {
```

Comment:
There might be other whitespace characters we should look for. Digging into the [history of this repo](https://github.com/eclipse-uprotocol/up-cpp/blob/up-v1.5.7/include/up-cpp/uri/tools/Utils.h#L51), this check used to be done using [`std::isspace()`](https://en.cppreference.com/w/cpp/locale/isspace). There's also [`std::isblank()`](https://en.cppreference.com/w/cpp/locale/isblank), which _might_ be a better choice (I'm actually not entirely clear the difference).

Reply:
I rewrote it to use isspace (catches all space type characters, tab, newline, carriage return, etc).